src/Raven.Server/Documents/Queries/Graph/QueryQueryStep.cs

```diff
@@ -62,6 +62,9 @@ public bool CanBeConsideredForDestinationOptimization
         {
             get
             {
+                if (HasWhereClause) //TODO: verify with Tal how good is this change
```

Comment:
This is wrong, because collection queries should work, the problem is probably that we aren't filtering the ids.

Reply:
In the original code we used HasWhereClause and indeed it is needed for id queries since they are collection queries and have a where clause.
So the quick fix is to add HasWhereClause to the other conditions.
If we want to support id queries than we need to support it in the IsMatchedBy method and see if the asked id is matched by the where clause in the CollectionDestinationStep